text/0000-de-rfc-type-ascription.md

```diff
@@ -0,0 +1,104 @@
+
+This RFC intends to advocate for the feature being removed entirely (or at least moving the implementation to a less prime area of the syntactic space) with a fresh RFC being necessary to add it again.
+
+# Demotivation
+[demotivation]: #demotivation
+
+One of the primary demotivations is the negative effect on diagnostics. `:` is a pretty highly used syntactic component in Rust: a double colon is a crucial part of how paths work, and paths are found _everywhere_ in Rust (every variable and type name is a path). It's a rather easy to accidentally type `:` instead of `::`, and that is often interpreted as type ascription syntax. It's a terrible experience for the user to make an often hard-to-notice mistake and get told they are trying to use a feature they may not even have heard of. Here's [an example][ekuber-tweet] of a bad diagnostic caused by type ascription; while such diagnostics can be fixed, there are just so many of them. The fact that this is _still_ a problem despite the amazing work being put into diagnostics is a signal that it may always be a problem. And if this feature were to stabilize as-is, it would likely get worse since there would be backpressure to improve the diagnostics of legitimate uses of ascription. Good diagnostics are an exercise in guessing user intent; and the harder we make that the worse the diagnostics will be.
+
+The other demotivation is a shift of syntactic norms.
+
+Type ascription was originally RFCd in 2015. This is a time before `?`, an RFC that was _extremely_ controversial at its time but is now considered a very normal and sensible feature by the community. Similarly, while it may be harder to make that same exact claim about `.await`, the community has definitely softened on it since it was originally proposed. If type ascription were proposed today, it seems unlikely that the syntax would be chosen to be what it is now.
+
+Syntax isn't the only reason; while type ascription is probably a good idea, a feature this significant deserves to be properly designed for the zeitgeist.
+
+# Guide-level obfuscation
+[guide-level-obfuscation]: #guide-level-obfuscation
+
+The `:` type ascription syntax would be removed from the nightly language. It is up to the compiler team whether they wish to remove it completely from the compiler (or perhaps just make it unparseable and use some magical unstable `ascript!()` macro in the meantime so that it is testable).
+
+This does not prevent future type ascription RFCs from happening, however they must propose the feature from first principles, and justify their choice of syntax. They are, of course, free to copy the work or text of the previous RFC.
+
+# Reference-level obfuscation
+[reference-level-obfuscation]: #reference-level-obfuscation
+
+![](https://user-images.githubusercontent.com/1617736/187055431-2ab9f46b-4c23-4ec4-9884-d050501bf0c2.png)
```

Comment:
Let us not use images in RFC text (they aren’t always readable), or at least take care to provide the alt-text.

```suggestion
![diff shortstat of 275k removed lines](https://user-images.githubusercontent.com/1617736/187055431-2ab9f46b-4c23-4ec4-9884-d050501bf0c2.png)
```